accord-core/src/main/java/accord/local/durability/DurabilityLevel.java

```diff
@@ -0,0 +1,140 @@
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package accord.local.durability;
+
+import java.util.Objects;
+import javax.annotation.Nullable;
+
+import accord.local.Node;
+import accord.local.durability.DurabilityService.SyncLocal;
+import accord.local.durability.DurabilityService.SyncRemote;
+import accord.utils.SortedArrays.SortedArrayList;
+
+import static accord.local.durability.DurabilityService.SyncLocal.NoLocal;
+import static accord.local.durability.DurabilityService.SyncRemote.NoRemote;
+
+public class DurabilityLevel
+{
+    public static final DurabilityLevel NONE = new DurabilityLevel(NoLocal, NoRemote, null, null);
+
+    public final SyncLocal local;
+    public final SyncRemote remote;
+    public final @Nullable SortedArrayList<Node.Id> including;
+    public final @Nullable SortedArrayList<Node.Id> excluding;
+
+    public DurabilityLevel(SyncLocal local, SyncRemote remote, @Nullable SortedArrayList<Node.Id> including)
+    {
+        this(local, remote, including, null);
+    }
+
+    public DurabilityLevel(SyncLocal local, SyncRemote remote, @Nullable SortedArrayList<Node.Id> including, @Nullable SortedArrayList<Node.Id> excluding)
+    {
+        this.local = local;
+        this.remote = remote;
+        this.including = including;
+        this.excluding = excluding;
+    }
+
+    public boolean equals(Object that)
+    {
+        return that instanceof DurabilityLevel && equals((DurabilityLevel) that);
+    }
+
+    private boolean equals(DurabilityLevel that)
+    {
+        return this.local == that.local
+               && this.remote == that.remote
+               && Objects.equals(this.including, that.including)
+               && Objects.equals(this.excluding, that.excluding);
+    }
+
+    @Override
+    public String toString()
+    {
+        return "{" +
+               "local=" + local +
+               ", remote=" + remote +
+               ", including=" + including +
+               ", excluding=" + excluding +
+               '}';
+    }
+
+    public static DurabilityLevel min(DurabilityLevel a, DurabilityLevel b)
+    {
+        SyncLocal local = min(a.local, b.local);
+        SyncRemote remote = min(a.remote, b.remote);
+        SortedArrayList<Node.Id> including = union(a.including, b.including);
```

Comment:
I might be misunderstanding this: if we want to find out a minimum durability level, don't we want to subtract `b.including` from `a.including` (i.e. find an intersection)?

Reply:
including/excluding are opposites, and we union both then subtract excluding from the combination of including. The reason to do it this way is we might contact different sets of nodes, so have included in one round but not another doesn't mean we haven't succeeded on that node. But if the round contacted the node but had no reply, it will be included in the `excluding` collection